### ISSUE TITLE
YCM 0.10 and CONFIGURE_SOURCE_DIR vs SOURCE_SUBDIR

### ISSUE THREAD

**Issue by PeterBowman** [robotology/ycm#210]
I used to take advantage of the `CONFIGURE_SOURCE_DIR` option to `ycm_ep_helper()`, further propagated to the ExternalProject customized copy in cmake-next/proposed. This option is still handled in `ycm_ep_helper()` after https://github.com/robotology/ycm/commit/ae7e854a593bf13f48c67ecd3af4b2b23d79bd9c (see [this line](https://github.com/robotology/ycm/blob/1a231f296f859b4e476707fd655e5c5df7c98ce4/modules/YCMEPHelper.cmake#L918)), but not propagated anymore to ExternalProject. Affects current devel branch (YCM 0.10).

I made my own copy of Buildorocos_kdl.cmake based on [codyco-superbuild](https://github.com/robotology/codyco-superbuild/blob/a1a8794bff7348bc1ba5f2f9099e203fc1d3c29b/cmake/Buildorocos_kdl.cmake). Is this working for you on latest YCM devel?

**Comment by traversaro** [comment]
> I made my own copy of Buildorocos_kdl.cmake based on codyco-superbuild. Is this working for you on latest YCM devel?

We are currently deprecating the codyco-superbuild and we do not have the `Buildorocos_kdl.cmake` in the robotology-superbuild , so I am afraid that is untested.

**Comment by drdanz** [comment]
It is propagated, see

https://github.com/robotology/ycm/blob/1a231f296f859b4e476707fd655e5c5df7c98ce4/modules/YCMEPHelper.cmake#L913-L919

https://github.com/robotology/ycm/blob/1a231f296f859b4e476707fd655e5c5df7c98ce4/modules/YCMEPHelper.cmake#L990-L1000

I think that this used to be an extension of `ExternalProject`, I don't remember exactly (I don't even remember adding it, and git history is not helping), but I probably considered it fixed when I saw the [`SOURCE_SUBDIR`](https://github.com/Kitware/CMake/commit/a8345d65f359d75efb057d22976cfb92b4d477cf#diff-05dc5703e53e101b74fd85c8a728625f)

I think to keep it coherent, we should add an option `SOURCE_SUBDIR`, and deprecate `CONFIGURE_SOURCE_DIR`, mapping it to source subdir.

**Comment by PeterBowman** [comment]
I'm a bit baffled by how CMake treats local vs cached variables. I'm following [this example](http://robotology.github.io/ycm/gh-pages/git-master/manual/ycm-superbuild-example.7.html#superbuild-example-developer-point-of-view) regarding the required (for my use case) `YCM_USE_CMAKE_PROPOSED` var. In short, it boils down to:

```cmake
set(YCM_USE_CMAKE_PROPOSED ON)
include(YCMBootstrap)
include(FindOrBuildPackage)
# do stuff, i.e. request Buildorocos_kdl.cmake
```

Relevant lines on the YCM side:

https://github.com/robotology/ycm/blob/1a231f296f859b4e476707fd655e5c5df7c98ce4/YCMConfig.cmake.in#L44-L48

This code is not taking into account the proposed module path after the first configure execution. My `set(YCM_USE_CMAKE_PROPOSED ON)` instruction only takes effect on the second run and onwards, which might have misled me in my previous experiments.

Can anyone reproduce this behaviour? Should the linked documentation be updated?

**Comment by drdanz** [comment]
Hmm I think you are right. I tested this with every CMake version since 2.8.7:

```cmake
cmake_minimum_required(VERSION 2.8.7)
project(test_cmake_option NONE)

set(FOO ON)
option(FOO "Foo" OFF)

set(BAR ON)
set(BAR OFF CACHE BOOL "Bar")

message("FOO = ${FOO}")
message("BAR = ${BAR}")
```

The output is always `OFF` the first time and `ON` the second. I don't know if this was different in previous CMake versions, but 2.8.7 is old enough to let me think that that this never worked on the first execution.

Anyway, in my opinion the real problem is the documentation, but the fact that `YCM_USE_CMAKE_PROPOSED` is cached... See also #57 (`YCM_USE_CMAKE_PROPOSED` was probably added after opening that issue, therefore it is not listed there).

**Comment by PeterBowman** [comment]
Thanks for taking a look at that. It strikes me that such behaviour is somewhat omitted in CMake docs ([v3.0](https://cmake.org/cmake/help/v3.0/command/set.html), [v3.14](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#cmake-language-variables)), or at least it's not clear enough to me.

To put this in context, I'm setting said YCM variable [here](https://github.com/roboticslab-uc3m/asibot-main/blob/0259f65fb5c7043319ca0f8d0cc212f15d171b09/CMakeLists.txt#L15-L17). For now, I'd force it to `ON` in the cache: `set(YCM_USE_CMAKE_PROPOSED ON CACHE BOOL "" FORCE)`.

**Comment by drdanz** [comment]
Fixed by #211

**Comment by drdanz** [comment]
@PeterBowman I opened #212 to track the `CACHE` issue on a separate issue
